server/app/durablejobs/DurableJobRunner.java

```diff
@@ -0,0 +1,178 @@
+  }
+
+  private void runJob(PersistedDurableJob persistedDurableJob) {
+    LocalDateTime startTime = nowProvider.get();
+    LOGGER.info(
+        "JobRunner_ExecutingJob thread_ID={}, job_name=\"{}\", job_ID={}",
+        Thread.currentThread().getId(),
+        persistedDurableJob.getJobName(),
+        persistedDurableJob.id);
+
+    try {
+      persistedDurableJob.decrementRemainingAttempts().save();
+
+      // Run the job in a separate thread and block until it completes, fails, or times out.
+      runJobWithTimeout(
+          durableJobRegistry
+              .get(DurableJobName.valueOf(persistedDurableJob.getJobName()))
+              .getFactory()
+              .create(persistedDurableJob));
+
+      persistedDurableJob.setSuccessTime(nowProvider.get().toInstant(zoneOffset)).save();
+
+      LOGGER.info(
+          "JobRunner_JobSucceeded job_name=\"{}\", job_ID={}, duration_s={}",
+          persistedDurableJob.getJobName(),
+          persistedDurableJob.id,
+          getJobDurationInSeconds(startTime));
+    } catch (JobNotFoundException
+        | IllegalArgumentException
+        | CancellationException
+        | InterruptedException e) {
+      String msg =
+          String.format(
+              "JobRunner_JobFailed %s job_name=\"%s\", job_ID=%d, attempts_remaining=%d,"
+                  + " duration_s=%f",
+              e.getClass().getSimpleName(),
+              persistedDurableJob.getJobName(),
+              persistedDurableJob.id,
+              persistedDurableJob.getRemainingAttempts(),
+              getJobDurationInSeconds(startTime));
+      LOGGER.error(msg);
+      persistedDurableJob.appendErrorMessage(msg).save();
+    } catch (TimeoutException e) {
+      String msg =
+          String.format(
+              "JobRunner_JobTimeout job_name=\"%s\", job_ID=%d, attempts_remaining=%d,"
+                  + " duration_s=%f",
+              persistedDurableJob.getJobName(),
+              persistedDurableJob.id,
+              persistedDurableJob.getRemainingAttempts(),
+              getJobDurationInSeconds(startTime));
+      LOGGER.error(msg);
+      persistedDurableJob.appendErrorMessage(msg).save();
+    } catch (ExecutionException e) {
+      String msg =
+          String.format(
+              "JobRunner_JobFailed ExecutionException job_name=\"%s\", job_ID=%d,"
+                  + " attempts_remaining=%d, duration_s=%f, error_message=%s, trace=%s",
+              persistedDurableJob.getJobName(),
+              persistedDurableJob.id,
+              persistedDurableJob.getRemainingAttempts(),
+              getJobDurationInSeconds(startTime),
+              e.getMessage(),
+              ExceptionUtils.getStackTrace(e));
+      LOGGER.error(msg);
+      persistedDurableJob.appendErrorMessage(msg).save();
+    }
+  }
+
+  private LocalDateTime resolveStopTime() {
+    // We set poll interval to 0 in test
+    if (runnerLifespanSeconds == 0) {
```

Comment:
For this and the `jobTimeoutMinutes` below, would it make sense to set the timeout in tests, rather than putting this in the code to cover tests?  I just wonder if we should throw an exception if the value given is <= 0, rather than effectively making 0 == 5.  Or perhaps disable the timeout entirely if it's 0?

Reply:
It's useful to test that the timeout works, so I'd rather not disable timeout entirely.

The tricky bit is that we have to specify chrono units with the time, making passing much shorter times in for testing awkward

Reply:
I'd also prefer not to have the special case in the code but I'm not seeing a way to do it that does all of:
- gives us a short timeout for testing
- doesn't require passing in an awkward number to the server config e.g. `600` for specifying a 10 minute timeout